kernel/src/fs/cgroupfs/controller/mod.rs

```diff
@@ -0,0 +1,431 @@
+        &self,
+        name: &str,
+        reader: &mut VmReader,
+        cgroup_node: &dyn CgroupSysNode,
+    ) -> Result<usize> {
+        self.as_subcontrol().write_attr(name, reader, cgroup_node)
+    }
+}
+
+bitflags! {
+    /// Bitflags representing active/deactive sub-control state.
+    pub(super) struct SubCtrlState: u8 {
+        const MEMORY_CTRLS = 1 << 0;
+        const CPUSET_CTRLS = 1 << 1;
+        const PIDS_CTRLS = 1 << 2;
+    }
+}
+
+impl SubCtrlState {
+    pub(super) fn control_bit(ctrl_type: ControllerType) -> Self {
+        match ctrl_type {
+            ControllerType::Memory => Self::MEMORY_CTRLS,
+            ControllerType::CpuSet => Self::CPUSET_CTRLS,
+            ControllerType::Pids => Self::PIDS_CTRLS,
+        }
+    }
+
+    /// Checks if a sub-control is active in the current state.
+    ///
+    /// If the given name does not represent a supported controller,
+    /// returns `None`.
+    pub(super) fn is_active(&self, ctrl_type: ControllerType) -> bool {
+        self.contains(Self::control_bit(ctrl_type))
+    }
+
+    fn activate(&mut self, ctrl_type: ControllerType) {
+        *self |= Self::control_bit(ctrl_type)
+    }
+
+    fn deactivate(&mut self, ctrl_type: ControllerType) {
+        *self -= Self::control_bit(ctrl_type)
+    }
+
+    pub(super) fn show(&self) -> String {
+        let mut controllers = Vec::new();
+
+        if self.contains(Self::MEMORY_CTRLS) {
+            controllers.push("memory");
+        }
+        if self.contains(Self::CPUSET_CTRLS) {
+            controllers.push("cpuset");
+        }
+        if self.contains(Self::PIDS_CTRLS) {
+            controllers.push("pids");
+        }
+
+        controllers.join(" ")
+    }
+
+    /// Returns an iterator over the active controller types.
+    pub(super) fn iter_active(&self) -> impl Iterator<Item = ControllerType> + '_ {
+        ControllerType::ALL
+            .into_iter()
+            .filter(|&ctrl_type| self.is_active(ctrl_type))
+    }
+}
+
+/// The controller for a single cgroup.
+///
+/// This struct can manage the activation state of each sub-control, and dispatches read/write
+/// operations to the appropriate sub-controllers.
+///
+/// The following is an explanation of the activation for sub-controls and controllers.
+/// When a cgroup activates a specific sub-control (e.g., memory, io), it means this control
+/// capability is being delegated to its children. Consequently, the corresponding controller
+/// within the child nodes will be activated.
+///
+/// The root node serves as the origin for all these control capabilities, so the controllers
+/// it possesses are always active. For any other node, only if its parent node first enables
+/// a sub-control, its corresponding controller will be activated.
+///
+/// Among all nodes, the fundamental cgroup controller is always active.
+pub(super) struct Controller {
+    sub_ctrl_state: Mutex<SubCtrlState>,
+
+    memory: RcuOption<Arc<SubController>>,
```

Comment:
So what about:
```
    memory: RcuOption<Arc<MemoryController>>,
```

I didn't realize there were fundamental problems preventing that.